File: src/utils/eventlistener.ts
export interface Eventlistener {
    type: string;
    callback: Function;
}

export class EventEmitter {
    protected listeners: Eventlistener[] = [];

    /**
     * Add event listener.
     */
    public addEventListener(type: string, callback: Function): Eventlistener {
        const listener: Eventlistener = { type, callback };
        this.listeners.push(listener);
        return listener;
    }

    /**
     * Remove event listener.
     * @param listener - Event listener to remove.
     */
    public removeEventListener(listener: unknown) {
        for (let c = 0; c < this.listeners.length; c++) {
            if (listener === this.listeners[c]) {
                this.listeners.splice(c, 1);
                return;
            }
        }
    }

    /**
     * Remove event listeners.
     * @param listeners - List of event listeners to remove.
     */
    public removeEventListeners(listeners: Eventlistener[]) {
        listeners.forEach((listener: Eventlistener) => {
            this.removeEventListener(listener);
        });
    }

    /**
     * Trigger event.
     * @param ce - Custom event to dispatch.
     */
    public dispatchEvent(ce: Event) {
        const listeners = this.listeners.slice();
        listeners.forEach(function (l) {
            if (ce.type === l.type) {
                //@ts-ignore
                l.callback.apply(this, [ce]);
            }
        });
    }
}

File: src/python/eventlistener.py
from typing import List, Callable, Any
from dataclasses import dataclass


@dataclass
class EventListener:
    type: str
    callback: Callable


class EventEmitter:
    def __init__(self):
        self.listeners: List[EventListener] = []

    def add_event_listener(self, event_type: str, callback: Callable) -> EventListener:
        """Add event listener."""
        listener = EventListener(type=event_type, callback=callback)
        self.listeners.append(listener)
        return listener

    def remove_event_listener(self, listener: EventListener) -> None:
        """Remove event listener."""
        if listener in self.listeners:
            self.listeners.remove(listener)

    def remove_event_listeners(self, listeners: List[EventListener]) -> None:
        """Remove event listeners."""
        for listener in listeners:
            self.remove_event_listener(listener)

    def dispatch_event(self, event: Any) -> None:
        """Trigger event."""
        listeners = self.listeners.copy()
        for listener in listeners:
            if event.type == listener.type:
                listener.callback(event)

